index.bs

```diff
@@ -3552,73 +3552,49 @@ Methods</h4>
-
-			4. If \(E_1\) exists:
-				1. If \(E_1\) is a <code>setTarget</code> event,
-					1. Implicitly insert a <code>setValueAtTime</code>
-						event at time \(t_c\) with the value that the
-						<code>setTarget</code> would have at time
-						\(t_c\).
-						<img alt="Graphical representation of calling cancelAndHoldAtTime when setTargetAtTime has been called at this time" src="images/cancel-setTarget.svg" style="height:75%;width:75%">
-
-					2. Go to step 5.
-
-				2. If \(E_1\) is a <code>setValueCurve</code> with a start
-					time of \(t_3\) and a duration of \(d\)
-
-					1. If \(t_c \gt t_3 + d\), go to step 5.
-
-					2. Otherwise,
-						1. Effectively replace this event with a
-							<code>setValueCurve</code> event with a start time
-							of \(t_3\) and a new duration of \(t_c-t_3\).
-							However, this is not a true replacement; this
-							automation MUST take care to produce the same
-							output as the original, and not one computed using
-							a different duration. (That would cause sampling of
-							the value curve in a slightly different way,
-							producing different results.)
-							<img alt="Graphical representation of calling cancelAndHoldAtTime when setValueCurve has been called at this time" src="images/cancel-setValueCurve.svg" style="height:75%;width:75%">
-
-						2. Go to step 5.
-
-			5. Remove all events with time greater than \(t_c\).
-
-			If no events are added, then the automation value after
-			{{AudioParam/cancelAndHoldAtTime()}} is the constant value that
-			the original timeline would have had at time \(t_c\).
-		</div>
+		When <code>cancelAndHoldAtTime</code> is called with a time <var>t</var> as
+		a parameter, execute the following steps:
+
+		- Let <var>v</var> be the value of the {{AudioParam}} at time
+			<var>t</var>.
+		- Let <var>current</var> be the <a lt="current audioparam">current
+			AudioParam</a> at time <var>t</var>.
+		- <a lt="rewrite event">Rewrite the current AudioParam event</a> to
+			end at <var>t</var>.
+		- Remove all {{AudioParam}} event that are in the list after
+			<var>current</var>.
+		- Insert a new <code>setValueAtTime</code> event at time <var>t</var> with
+			value <var>v</var>.
+
+		The <dfn dfn-type="dfn" lt="current audioparam">current AudioParam</dfn> at time
+		<var>t</var> is the {{AudioParam}} that is being used to compute the
+		value of the {{AudioParam}} at time <var>t</var>.
+
+		Note: There is only one [=current AudioParam=] for a givent time
+			<var>t</var>. Depending on the {{AudioParam}} event type, the time of the
+			event itself can be greater or lower than <var>t</var>.
+
+		<dfn dfn-type="dfn" lt="rewrite event">Rewriting a AudioParam event</dfn> to end at
+		time <var>t</var> means changing the {{AudioParam}} event in a way that:
+		- The values for this {{AudioParam}} before and at <var>t</var> are not
+			changed.
+		- If this {{AudioParam}} <a>has an end time</a>, it now ends at
+			<var>t</var>.
+
```

Comment:
It is necessary to say what happens if there is no current automation event at time t. Some phrasing like "If there is a current automation event at time t, rewrite it to end at t" should suffice.

Also someone has to decide if in the case of no automation event at time t, the setValueAtTime should still be called at t (this has consequences for subsequent ramp start time). I think I'm in favor of always calling setValueAtTime -- more predictable...

Reply:
I don't understand:

- There is no situation where there is no current automation event. Worst case it's the initial value for this `AudioParam` at time 0.
- The prose that inserts a `setValueAtTime` event is not in any sort of conditional, so it's always going to be inserted, so ramps, etc. will always start at `t`.

Reply:
Does initial value qualify as an automation event, though? I don't think the spec says so.
(BTW, I like the draft version you wrote – nice and simple. The above is just a formal nitpick. Feel free to ignore if you feel it's unimportant.)

Reply:
Yes. I think it should. Perhaps what we want is to have an initial event of `setValueAtTime(defaultValue,0)` in the event list. I thought we had that in the spec, but I couldn't find it.